### PR TITLE
Implemented eventarc event publishing API (#1617)

### PR DESCRIPTION
**Pull request by pavelgj**

### CODE REVIEW

File: src/eventarc/eventarc.ts [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+   * ```javascript
+   * var app = channel.eventarc;
+   * ```
+   */
+  get eventarc(): Eventarc {
+    return this.eventarcInternal;
+  }
+
+  /**
+   * The channel name as provided during channel creation except if it was not specifed then the default
```
Comment (egilmorez) [inline src/eventarc/eventarc.ts:175]: Suggest a stop, like "was not specified; then the..."

(note typo as well)

**Review by egilmorez — COMMENTED** [review]
A few things to look at Pavel, thanks!

Kevin should also be on this PR.  I *think* he'll write all the guidance for this, but I'm still not sure.

File: src/eventarc/cloudevent.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+/**
+  * A CloudEvent describes event data.
+  * 
+  * @see https://github.com/cloudevents/spec/blob/v1.0/spec.md
+  */
+export interface CloudEvent {
+
+   /** 
+    * Event identified. If not provided will be auto-generated with a UUID.
```
[the remark this answers is not in the document]
Comment (pavelgj) [reply src/eventarc/cloudevent.ts]: Done.

File: src/eventarc/cloudevent.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    * 
+    * @see https://github.com/cloudevents/spec/blob/v1.0/spec.md#id 
+    */
+   id?: string;
+
+   /**
+    * Identifies the context in which an event happened. If not provided the value of `EVENTARC_CLOUD_EVENT_SOURCE`
```
[the remark this answers is not in the document]
Comment (pavelgj) [reply src/eventarc/cloudevent.ts]: Done.

File: src/eventarc/cloudevent.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+   /**
+    * Subject (context) of the event in the context of the event producer.
+    * 
+    * @see https://github.com/cloudevents/spec/blob/v1.0/spec.md#subject
+    */
+   subject?: string;
+
+   /**
+    * MIME type of the data being sent with the event in the `data` field. Only `application/json` and `text/plain`
+    * are currently supported. If not specified wil be automatically inferred from the type of provided data.
```
[the remark this answers is not in the document]
Comment (pavelgj) [reply src/eventarc/cloudevent.ts]: Done.

File: src/eventarc/cloudevent.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    * 
+    * @see https://github.com/cloudevents/spec/blob/v1.0/spec.md#datacontenttype
+    */
+   datacontenttype?: string;
+
+   /**
+    * Timestamp of when the occurrence happened. Must in ISO time format. If not specified current time (at the
```
[the remark this answers is not in the document]
Comment (pavelgj) [reply src/eventarc/cloudevent.ts]: Done.

File: src/eventarc/eventarc.ts [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+        'invalid-argument',
+        'First argument passed to Eventarc() must be a valid Firebase app instance.',
+      );
+    }
+
+    this.appInternal = app;
+  }
+
+  /**
+   * The {@link firebase-admin.app#App} associated with the current `Eventarc` service
```
[the remark this answers is not in the document]
Comment (pavelgj) [reply src/eventarc/eventarc.ts]: removed backtick, here it refers to the service name.

File: src/eventarc/eventarc.ts [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+   * var app = eventarc.app;
+   * ```
+   */
+  get app(): App {
+    return this.appInternal;
+  }
+
+  /**
+   * Creates a reference to the Eventarc channel using the provided channel resource name.
+   * The channel resource name Can be either:
```
[the remark this answers is not in the document]
Comment (pavelgj) [reply src/eventarc/eventarc.ts]: Done.

File: src/eventarc/eventarc.ts [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+   *   * fully qualified channel resource name:
+   *     `projects/{project}/locations/{location}/channels/{channel-id}`
+   *   * partial resource name with location and channel ID, in which case
+   *     the runtime project ID of the function will be used:
+   *     `locations/{location}/channels/{channel-id}`
+   *   * partial channel-id, in which case the runtime project ID of the
```
[the remark this answers is not in the document]
Comment (pavelgj) [reply src/eventarc/eventarc.ts]: done.

File: src/eventarc/eventarc.ts (hunk 1 above)
Comment (pavelgj) [reply src/eventarc/eventarc.ts]: Done.

**Comment by egilmorez** [comment]
Thanks Pavel!  

Could you take a pass through for the two global nits?  This was the "ask:"

Wherever possible, use present tense is instead of "will" or "will be."

Also, commas after phrases like "if not provided" help the flow.

There are several more instances of both to look at.

**Comment by pavelgj** [comment]
@egilmorez made another global pass as per your suggestion. PTAL.

File: src/eventarc/eventarc.ts [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+   *   * partial resource name with location and channel ID, in which case
+   *     the runtime project ID of the function is used:
+   *     `locations/{location}/channels/{channel-id}`
+   *   * partial channel ID, in which case the runtime project ID of the
+   *     function and `us-central1` as location is used:
+   *     `{channel-id}`
+   * 
+   * @param name - Channel resource name. 
+   * @param options - (optional) additional channel options
+   * @returns Eventarc channel reference for publishing events.
```
Comment (kevinthecheung) [inline src/eventarc/eventarc.ts:84]:
```suggestion
   * Creates a reference to the Eventarc channel using the provided channel resource name.
   * The channel resource name can be either:
   *   * A fully qualified channel resource name:
   *     `projects/{project}/locations/{location}/channels/{channel-id}`
   *   * A partial resource name with location and channel ID, in which case
   *     the runtime project ID of the function is used:
   *     `locations/{location}/channels/{channel-id}`
   *   * A partial channel ID, in which case the runtime project ID of the
   *     function and `us-central1` as location is used:
   *     `{channel-id}`
   * 
   * @param name - Channel resource name. 
   * @param options - (optional) additional channel options
   * @returns An Eventarc channel reference for publishing events.
```

File: src/eventarc/eventarc.ts [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+
+/**
+ * Eventarc Channel.
+ */
+export class Channel {
+  private readonly eventarcInternal: Eventarc;
+  private nameInternal: string;
+
+  /**
+   * List if event types allowed by this channel for publishing. Other event types are ignored.
```
Comment (kevinthecheung) [inline src/eventarc/eventarc.ts:134]:
```suggestion
   * List of event types allowed by this channel for publishing. Other event types are ignored.
```

**Review by kevinthecheung — APPROVED** [review]

File: src/eventarc/cloudevent.ts [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+   /**
+    * Timestamp of the event. Must be in ISO time format. If not specified, current time (at the moment of publishing)
+    * is used.
+    * 
+    * @see https://github.com/cloudevents/spec/blob/v1.0/spec.md#time
+    */
+   time?: string;
+
+   /**
+    * Data payload of the event. Objects are strigified with JSON and strings are be passed along as-is.
```
Comment (egilmorez) [inline src/eventarc/cloudevent.ts:85]: Typo on "stringified"

**Review by egilmorez — APPROVED** [review]
Found one typo Pavel (and a "will," but I'm not here to be *that* annoying).

Thanks!

File: src/eventarc/cloudevent.ts (hunk 11 above)
Comment (pavelgj) [reply src/eventarc/cloudevent.ts]: Done. Also fixed remaining wills. :)

**Review by lahirumaramba — APPROVED** [review]

**Pull request merged**